util/kube/kube.go

```diff
@@ -6,11 +6,18 @@ import (
 	"k8s.io/apimachinery/pkg/runtime/schema"
 
 	"github.com/argoproj/argo-cd/common"
+	"github.com/argoproj/argo-cd/util/settings"
 )
 
 // SetAppInstanceLabel the recommended app.kubernetes.io/instance label against an unstructured object
 // Uses the legacy labeling if environment variable is set
 func SetAppInstanceLabel(target *unstructured.Unstructured, key, val string) error {
+	// if val is larger than "max length" long, set annotation instead - see #4045
+	// only do this if not {apps, extensions}/v1beta1 - see #335 and special handling below
+	if len(val) >= settings.AppInstanceLabelMaxLen && target.GetAPIVersion() != "apps/v1beta1" && target.GetAPIVersion() != "extensions/v1beta1" {
```

Comment:
Annotation is preferable in some other cases, not only when the application name is too long. 

Instead of automatically switching to annotation let's introduce an application-level setting that allows switching from label to annotation. I would suggest using the `argocd.argoproj.io/annotation-marker` annotation . E.g. if application has `argocd.argoproj.io/annotation-marker` annotation then argocd should make resources with annotation:


```
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: guestbook
  annotations:
    # instruct argocd to use annotation instead of label
    argocd.argoproj.io/annotation-marker: 'true'
```